com.spotify/index.js

```diff
@@ -0,0 +1,662 @@
+        'user-read-playback-state',
+        'user-read-currently-playing',
+        'user-modify-playback-state',
+        'user-read-recently-played',
+        'user-top-read',
+        'user-follow-read',
+        'user-follow-modify',
+        'user-library-read',
+        'user-library-modify'],
+      set_state: true,
+      authorize: 'https://accounts.spotify.com/authorize',
+      get_access_token: 'https://accounts.spotify.com/api/token',
+      redirect_uri: platform.getOrigin() + '/devices/oauth2/callback/com.spotify',
+
+      callback(engine, accessToken, refreshToken) {
+        return Tp.Helpers.Http.get(USER_URL, {
+          accept: 'application/json',
+          extraHeaders: {
+            'Authorization': 'Bearer ' + accessToken
+
+          }
+        }).then(response => {
+          const parsed = JSON.parse(response);
+          console.log(parsed);
+          return engine.devices.loadOneDevice(
+            {
+              kind: 'com.spotify',
+              accessToken: accessToken,
+              refreshToken: refreshToken,
+              userId: parsed.id,
+              userName: parsed.display_name
+            },
+            true
+          );
+        });
+      }
+    });
+  }
+
+  constructor(engine, state) {
+    super(engine, state);
+
+    this.uniqueId = 'com.spotify-' + this.userId;
+    this.name = 'Spotify Account of %s'.format(this.userName);
+    this.description = 'This is your Spotify Account';
+  }
+
+  get userId() {
+    return this.state.userId;
+  }
+
+  get userName() {
+    return this.state.userName;
+  }
+
+  get_get_user_name() {
+    return [{username: this.state.userName}];
+  }
+
+  get_get_user_id() {
+    return [{id: this.state.userId}];
+  }
+
+
+  /* Player */
+
+  currently_playing_helper() {
+    return Tp.Helpers.Http.get(CURRENTLY_PLAYING_URL, {
+      accept: 'application/json',
+      useOAuth2: this
+      /*
+      // TODO can we replace this by setting useOAuth2: this ?
+      // most likely - see https://almond.stanford.edu/thingpedia/developers/thingpedia-helpers.md#module-helpers-http
+      extraHeaders: {
+        'Authorization': 'Bearer ' + this.state.accessToken
+      } */
+    });
+  }
+
+  get_get_currently_playing() {
+    return this.currently_playing_helper().then(response => {
+      let NONE_RESULT = [{name: '', is_song_playing: false,
+        human_output: "No song is currently playing."}];
+      if (response === '' || response.length === 0 ) {
+        return NONE_RESULT;
+      }
+      const parsed = JSON.parse(response);
+      if (parsed.is_playing == false) {
+        return NONE_RESULT;
+      }
+      let message = "The currently playing song is '" + parsed.item.name + "'.";
+      // TODO create convenience functions to parse out groups of artists, albums, and tracks
+      // TODO use the function to expand the message to include artists.
+      return [{name: parsed.item.name, is_song_playing: true, human_output: message}];
```

Comment:
The message should be added to the "formatted output" field of the manifest, not here.